pkg/reconciler/pullsubscription/pullsubscription.go

```diff
@@ -147,11 +149,43 @@ func (c *Reconciler) reconcile(ctx context.Context, source *v1alpha1.PullSubscri
 
 	if source.GetDeletionTimestamp() != nil {
 		logger.Info("Source Deleting.")
-		err := c.ensureSubscriptionRemoval(ctx, source)
-		if err != nil {
-			logger.Error("Unable to delete the Subscription", zap.Error(err))
+
+		state, err := c.EnsureSubscriptionDeleted(ctx, source, source.Spec.Project, source.Spec.Topic, source.Status.SubscriptionID)
+		switch state {
+		case pubsub.OpsGetFailedState:
+			logger.Error("Failed to get subscription ops job.", zap.Any("state", state), zap.Error(err))
+			return err
+
+		case pubsub.OpsCreatedState:
+			// If we created a job to make a subscription, then add the finalizer and update the status.
+			source.Status.MarkSubscriptionOperation(
+				"Deleting",
+				"Created Job to delete Subscription %q.",
+				source.Status.SubscriptionID)
+			return nil
+
+		case pubsub.OpsCompeteSuccessfulState:
+			source.Status.MarkNoSubscription(
+				"Deleted",
+				"Successfully deleted Subscription %q.",
+				source.Status.SubscriptionID)
+			source.Status.SubscriptionID = ""
+			removeFinalizer(source)
+
+		case pubsub.OpsCreateFailedState, pubsub.OpsCompeteFailedState:
```

Comment:
This leaves the finalizer hanging around -- should we give up on the finalizer at some point (for example, if we don't have permissions to delete the subscription, or if it is already deleted)?

I seem to recall that the GitHub source had issues where it would stick around forever if (for example) you created a source but it hadn't had a chance to register the webhook yet.

Reply:
you can edit the resource and remove the finalizer.